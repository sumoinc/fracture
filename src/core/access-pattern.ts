import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { FractureComponent } from "../core/component";
import { DynamoGsi } from "../dynamodb/dynamo-gsi";
import { DynamoTable } from "../dynamodb/dynamo-table";

export const ACCESS_PATTERN_TYPE = {
  /**
   * Main PK and PK pattern used to identify this record.
   */
  IDENTIFIER: "Identifier",
  /**
   * Lookup pattern for quick begins_with() searching and such.
   */
  LOOKUP: "lookup",
  /**
   * Any other AP
   */
  OTHER: "other",
} as const;

export interface AccessPatternOptions {
  dynamoGsi?: DynamoGsi;
  pkAttributeOptions?: ResourceAttributeOptions;
  skAttributeOptions?: ResourceAttributeOptions;
  type?: ValueOf<typeof ACCESS_PATTERN_TYPE>;
}

export class AccessPattern extends FractureComponent {
  // member components
  public pkAttribute: ResourceAttribute;
  public skAttribute: ResourceAttribute;
  public operations: Operation[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AccessPatternOptions>;

  constructor(resource: Resource, options: AccessPatternOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    // make a new gsi is one wasn't passed into the Access Pattern
    const dynamoGsi = options.dynamoGsi
      ? options.dynamoGsi
      : new DynamoGsi(resource.dynamoTable);

    const defaultOptions: Required<AccessPatternOptions> = {
      dynamoGsi,
      pkAttributeOptions: {
        name: dynamoGsi.pkName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      skAttributeOptions: {
        name: dynamoGsi.skName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      type: ACCESS_PATTERN_TYPE.OTHER,
    };

    /***************************************************************************
     *
     * INIT ACCESS PATTERN
     *
     **************************************************************************/

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<AccessPatternOptions>;
    // the attribute might already be defined.
    const pkAttribute = resource.getAttributeByName(dynamoGsi.pkName);
    const skAttribute = resource.getAttributeByName(dynamoGsi.skName);

    // member components
    this.pkAttribute = pkAttribute
      ? pkAttribute
      : new ResourceAttribute(resource, this.options.pkAttributeOptions);
    this.skAttribute = skAttribute
      ? skAttribute
      : new ResourceAttribute(resource, this.options.skAttributeOptions);

    // parent + inverse
    this.resource = resource;
    this.resource.accessPatterns.push(this);

    this.project.logger.info(
      `Access Pattern for: "${this.pkAttribute.name}:${this.skAttribute.name}" initialized.`
    );

    return this;
  }

  public build() {
    this.project.logger.debug(
      `BUILD Access Pattern: "${this.pkAttribute.name}:${this.skAttribute.name}" called.`
    );
    /***************************************************************************
     *
     * CRUD OPERATIONS
     *
     * TODO
     *
     **************************************************************************/
    /*
    new Operation(this, {
      operationType: OPERATION_TYPE.MUTATION,
      operationSubType: OPERATION_SUB_TYPE.LIST,
    });
    */
  }

  addPkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.pkAttribute.compositionSources.push(sourceAttribute);
  }

  addSkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.skAttribute.compositionSources.push(sourceAttribute);
  }

  /**
   * Returns operation for given sub-type
   */
  public getOperation(
    operationSubType: ValueOf<typeof OPERATION_SUB_TYPE>
  ): Operation {
    const returnOperation = this.operations.find(
      (o) => o.operationSubType === operationSubType
    );

    if (!returnOperation) {
      throw new Error(`Operation not found for sub-type: ${operationSubType}`);
    }

    return returnOperation;
  }

  public get createOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.CREATE_ONE);
  }

  public get readOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.READ_ONE);
  }

  public get updateOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.UPDATE_ONE);
  }

  public get deleteOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.DELETE_ONE);
  }

  public get isKeyAccessPattern(): boolean {
    return this.dynamoTable.keyDynamoGsi === this.dynamoGsi;
  }

  public get isLookupAccessPattern(): boolean {
    return this.dynamoTable.lookupDynamoGsi === this.dynamoGsi;
  }

  public get dynamoTable(): DynamoTable {
    return this.dynamoGsi.dynamoTable;
  }

  public get dynamoGsi(): DynamoGsi {
    return this.options.dynamoGsi;
  }
}
