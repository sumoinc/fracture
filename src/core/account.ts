import { Component } from "projen";
import { Organization } from "./organization";
import { OrganizationalUnit } from "./organizational-unit";
import { Region, RegionOptions } from "./region";

export interface AccountOptions {
  id: string;
  name?: string;
  rootEmail?: string;
  isManagementAccount?: boolean;
  organizationalUnit?: OrganizationalUnit;
}

export class Account extends Component {
  // member components
  public readonly regions: Region[];
  // parent
  public readonly organization: Organization;
  // all other options
  public readonly options: AccountOptions;

  constructor(organization: Organization, options: AccountOptions) {
    super(organization.project);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components
    this.regions = [];

    // parents + inverse
    this.organization = organization;
    this.organization.accounts.push(this);

    // all other options
    const defaultOptions: Partial<AccountOptions> = {
      isManagementAccount: false,
    };
    this.options = { ...defaultOptions, ...options };

    // if OU defined, add account to OU
    if (this.options.organizationalUnit) {
      this.options.organizationalUnit.accounts.push(this);
    }

    // debugging output
    this.project.logger.info(`INIT Account: "${this.id}"`);
  }

  public get id() {
    return this.options.id;
  }

  public get name() {
    return this.options.name;
  }

  public get rootEmail() {
    return this.options.rootEmail;
  }

  public get isManagementAccount() {
    return this.options.isManagementAccount;
  }

  public get organizationalUnit() {
    return this.options.organizationalUnit;
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add an region to an account.
   *
   * @param {RegionOptions}
   * @returns {Region}
   */
  public addRegion(options: RegionOptions) {
    const region = this.regions.find((r) => r.id === options.id);
    return region ? region : new Region(this, options);
  }
}
