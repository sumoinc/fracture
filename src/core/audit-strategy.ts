import {
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
  ResourceAttributeType,
} from "./resource-attribute";

export type AuditStrategy = {
  create: {
    dateAttribute?: ResourceAttributeOptions;
    userAttribute?: ResourceAttributeOptions;
  };
  update: {
    dateAttribute?: ResourceAttributeOptions;
    userAttribute?: ResourceAttributeOptions;
  };
  // turning this on activates soft deletes
  delete: {
    dateAttribute?: ResourceAttributeOptions;
    userAttribute?: ResourceAttributeOptions;
  };
};

export const defaultAuditStrategy: AuditStrategy = {
  create: {
    dateAttribute: {
      name: "created-at",
      shortName: "cd",
      comment: [`The date and time this record was created.`],
      type: ResourceAttributeType.DATE_TIME,
      isRequired: true,
      createGenerator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
    },
  },
  update: {
    dateAttribute: {
      name: "updated-at",
      shortName: "ud",
      comment: [`The date and time this record was last updated.`],
      type: ResourceAttributeType.DATE_TIME,
      isRequired: true,
      updateGenerator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
    },
  },
  delete: {
    dateAttribute: {
      name: "deleted-at",
      shortName: "dd",
      comment: [`The date and time this record was deleted.`],
      type: ResourceAttributeType.DATE_TIME,
      deleteGenerator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
    },
  },
};
