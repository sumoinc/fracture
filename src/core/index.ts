export * from "./access-pattern";
export * from "./account";
export * from "./audit-strategy";
export * from "./fracture";
export * from "./index";
export * from "./naming-strategy";
export * from "./operation";
export * from "./organization";
export * from "./organizational-unit";
export * from "./region";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure";
