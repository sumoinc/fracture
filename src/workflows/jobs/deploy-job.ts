import { paramCase } from "change-case";
import { Job, JobPermission, JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { SetOptional } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { Environment } from "../../environments";
import { renderDeploySteps } from "../steps/deploy-steps";
import { renderDownloadArtifactSteps } from "../steps/download-artifact-steps";
import { renderSetupNodeSteps } from "../steps/setup-node-steps";
import { Workflow } from "../workflow";

export interface DeployJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {
  /**
   * How to identify the app being deployed
   *
   */
  readonly appName: string;

  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPrefix?: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs?: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  readonly artifactsDirectory: string;
}

export class DeployJob extends WorkflowJob {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  public readonly branchPrefix: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  public readonly needs: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  public readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  public readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  public readonly artifactsDirectory: string;

  /**
   * Workflow for deployment
   */
  readonly workflow: Workflow;

  constructor(
    public readonly project: TypeScriptProject,
    options: DeployJobOptions
  ) {
    const branchPrefix = options.branchPrefix ?? "main";
    const name = `Deploy ${options.appName} to ${options.environment.name}`;
    const jobId = options.jobId ?? paramCase(name);

    super(project, {
      jobId,
      name,
      ...options,
    });

    // defaults
    this.workflow = Workflow.deploy(project);
    this.branchPrefix = branchPrefix;
    this.needs = [this.workflow.buildJob.jobId, ...(options.needs ?? [])];
    this.deploySteps = options.deploySteps;
    this.environment = options.environment;
    this.artifactsDirectory = options.artifactsDirectory;

    // add this branch pattern as a trigger for the workflow
    this.workflow.addPushTrigger(this.branchPrefix);

    // make sure build step will store the artifacts we need
    this.workflow.buildJob.artifactDirectories.push(this.artifactsDirectory);
  }

  dependsOn = (job: WorkflowJob): void => {
    this.needs.push(job.jobId);
  };

  render = (): Job => {
    return {
      name: this.name,
      runsOn: this.workflow.defaultRunners,
      needs: this.needs,
      concurrency: this.jobId,
      if: `startsWith( github.ref, 'refs/heads/${this.branchPrefix}' )`,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.READ,
        ...(this.environment.authProviderType === "AWS_GITHUB_OIDC"
          ? {
              idToken: JobPermission.WRITE,
            }
          : {}),
      },
      steps: [
        ...renderDownloadArtifactSteps(this),
        ...renderSetupNodeSteps(this.workflow, {
          packageCache: false,
          setupPnpm: false,
        }),
        ...renderDeploySteps(this),
      ],
    };
  };
}
