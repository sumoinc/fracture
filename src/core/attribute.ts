import { ValueOf } from "type-fest";
import { AccessPattern } from "./access-pattern";
import { FractureComponent } from "./component";
import { Resource } from "./resource";
import { Service } from "./service";

/**
 * Each Attribute has a type that is used to determine how we will construct other generated code.
 */
export const AttributeType = {
  /**
   *  A unique identifier for an object. This scalar is serialized like a String but isn't meant to be human-readable.
   *  Long format GUID
   */
  GUID: "GUID",
  /**
   *  In GraphQL:   A UTF-8 character sequence.
   *  In DynamoDB:  Maximum DynamoDB item size limit of 400 KB (DynamoDB limit)
   *                When used as DynamoDB Key:
   *                  For a simple primary key, the maximum length of the first attribute value (the partition key) is 2048 bytes.
   *                  For a composite primary key, the maximum length of the second attribute value (the sort key) is 1024 bytes.
   */
  STRING: "String",
  /**
   *  In GraphQL:   An integer value between -(231) and 232-1.
   *  In DynamoDB:  Stored as Number
   *                Numbers can be positive, negative, or zero. Numbers can have up to 38 digits of precision.
   *                Exceeding this results in an exception.
   *                Positive range: 1E-130 to 9.9999999999999999999999999999999999999E+125
   *                Negative range: -9.9999999999999999999999999999999999999E+125 to -1E-130
   */
  INT: "Int",
  /**
   *  In GraphQL:   An IEEE 754 floating point value.
   *  In DynamoDB:  Stored as Number
   *                Numbers can be positive, negative, or zero. Numbers can have up to 38 digits of precision.
   *                Exceeding this results in an exception.
   *                Positive range: 1E-130 to 9.9999999999999999999999999999999999999E+125
   *                Negative range: -9.9999999999999999999999999999999999999E+125 to -1E-130
   */
  FLOAT: "Float",
  /**
   *  In GraphQL:   A Boolean value, either true or false.
   */
  BOOLEAN: "Boolean",
  /*****************************************************************************
   *
   *  APPSYNC TYPES
   *
   *  Additional data types which AppSync supports.
   *
   ****************************************************************************/
  /**
   *  GraphQL:  An extended ISO 8601 date string in the format YYYY-MM-DD.
   *  DynamoDB: Stored as string
   */
  DATE: "Date",
  /**
   *  GraphQL:  An extended ISO 8601 time string in the format hh:mm:ss.sss.
   *  DynamoDB: Stored as string
   */
  TIME: "Time",
  /**
   *  GraphQL:  An extended ISO 8601 date and time string in the format YYYY-MM-DDThh:mm:ss.sssZ.
   *  DynamoDB: Stored as string
   */
  DATE_TIME: "DateTime",
  /**
   *  GraphQl:  An integer value representing the number of seconds before or after 1970-01-01-T00:00Z.
   *  DynamoDB: Stored as Number
   */
  TIMESTAMP: "Timestamp",
  /**
   *  GraphQL:  An email address in the format local-part@domain-part as defined by RFC 822.
   *  DynamoDB: Stored as string
   */
  EMAIL: "Email",
  /**
   *  GraphQL:  A JSON string. Any valid JSON construct is automatically parsed and loaded in the resolver mapping
   *            templates as maps, lists, or scalar values rather than as the literal input strings. Unquoted strings or
   *            otherwise invalid JSON result in a GraphQL validation error.
   *  DynamoDB: Stored as ???
   */
  JSON: "JSON",
  /**
   *  GraphQL:  A phone number. Phone numbers can contain either spaces or hyphens to separate digit groups. Phone
   *            numbers without a country code are assumed to be US/North American numbers adhering to the
   *            North American Numbering Plan (NANP).
   *  DynamoDB: Stored as string
   */
  PHONE: "Phone",
  /**
   *  GraphQL:  A URL as defined by RFC 1738. For example, https://www.amazon.com/dp/B000NZW3KC/ or
   *            mailto:example@example.com. URLs must contain a schema (http, mailto) and can't contain two forward
   *            slashes (//) in the path part.
   *  DynamoDB: Stored as string
   */
  URL: "URL",
  /**
   *  GraphQL:  A valid IPv4 or IPv6 address. IPv4 addresses are expected in quad-dotted notation (123.12.34.56).
   *            IPv6 addresses are expected in non-bracketed, colon-separated format (1a2b:3c4b::1234:4567). You can
   *            include an optional CIDR suffix (123.45.67.89/16) to indicate subnet mask.
   *  DynamoDB: Stored as string
   */
  IPADDRESS: "IPAddress",
  /*****************************************************************************
   *
   *  FRACTURE TYPES
   *
   *  Special types used by Fracture. That are currently used primarily to store values calculated based on other
   *  attributes.
   *
   ****************************************************************************/
  COUNT: "Count",
  AVERAGE: "Average",
  SUM: "Sum",
} as const;

export const AttributeGenerator = {
  AUTO_INCREMENT: "Increment",
  GUID: "Guid",
  CURRENT_DATE_TIME_STAMP: "CurrentDateTimeStamp",
  TYPE: "Type",
  NONE: "None",
  VERSION: "Version",
  /**
   * Composed of other attributes
   */
  COMPOSITION: "Composition",
} as const;

export const ValidationRule = {
  REQUIRED: "Required",
  TYPE: "Type",
} as const;

export type AttributeOptions = {
  /**
   * Full long name for this attribute.
   * @example 'phone-number'
   */
  name: string;
  /**
   * Brief name used when storing data to save space.
   * @example 'pn'
   * @default AttributeOptions.name
   */
  shortName?: string;
  /**
   * Comment lines to add to the Resource.
   * @default []
   */
  comments?: string[];
  /**
   * The type for this attribute.
   * @default AttributeType.STRING
   */
  type?: ValueOf<typeof AttributeType>;
  /**
   * Is this attribute required for all mutations?
   * @default false
   */
  isRequired?: boolean;
  /**
   * Is this attribute publically accessible / visible?
   * @default true
   */
  isPublic?: boolean;
  /**
   * Is this attribute used when building the PK?
   * @default false
   */
  isPkComponent?: boolean;
  /**
   * Is this attribute used when building the SK?
   * @default false
   */
  isSkComponent?: boolean;
  /**
   * Is this attribute used when lookuing up records?
   * @default false
   */
  isLookupComponent?: boolean;
  /**
   * The generator to use to build this attribute.
   * @default AttributeGenerator.NONE
   */
  generator?: ValueOf<typeof AttributeGenerator>;
  /**
   * Is this a attribute generated on create operations?
   * @default false
   */
  isGeneratedOnCreate?: boolean;
  /**
   * Is this a attribute generated on read operations?
   * @default false
   */
  isGeneratedOnRead?: boolean;
  /**
   * Is this a attribute generated on update operations?
   * @default false
   */
  isGeneratedOnUpdate?: boolean;
  /**
   * Is this a attribute generated on delete operations?
   * @default false
   */
  isGeneratedOnDelete?: boolean;
  /**
   * Validations to run when creating this attribute.
   * @default []
   */
  createValidations?: ValueOf<typeof ValidationRule>[];
  /**
   * Validations to run when reading data
   * @default []
   */
  readValidations?: ValueOf<typeof ValidationRule>[];
  /**
   * Validations to run when updating this attribute.
   * @default []
   */
  updateValidations?: ValueOf<typeof ValidationRule>[];
  /**
   * Validations to run when deleting this attribute.
   * @default []
   */
  deleteValidations?: ValueOf<typeof ValidationRule>[];
  /**
   * Validations to run when importing external records.
   * @default []
   */
  importValidations?: ValueOf<typeof ValidationRule>[];
  /**
   * Position this attribute should occupy when sorted.
   */
  sortPosition?: number;
  /**
   * The separator to use when composing this attribute from other attributes.
   * @default: '#'
   */
  compositionSeperator?: string;
};

export class Attribute extends FractureComponent {
  // member components
  public readonly compositionSources: Attribute[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AttributeOptions>;

  constructor(resource: Resource, options: Required<AttributeOptions>) {
    super(resource.fracture);
    this.resource = resource;
    this.options = options;
    return this;
  }

  /*****************************************************************************
   *
   * ACCESSORS
   *
   ****************************************************************************/

  public get isRequired(): boolean {
    return this.options.isRequired;
  }

  public get name(): string {
    return this.options.name;
  }

  public get shortName(): string {
    return this.options.shortName;
  }

  public get sortPosition(): number {
    const boost = this.isAccessPatternKey ? 1000 : 0;
    return this.options.sortPosition + boost;
  }

  /*****************************************************************************
   *
   * PK and SK HELPERS
   *
   ****************************************************************************/

  public get isPartitionKey(): boolean {
    return this.resource.partitionKey === this;
  }

  public get isSortKey(): boolean {
    return this.resource.sortKey === this;
  }

  /**
   * Is this key park of an access pattern?
   */
  public get isAccessPatternKey(): boolean {
    return this.resource.accessPatterns.some((accessPattern) => {
      return (
        accessPattern.pkAttribute.name === this.name ||
        accessPattern.skAttribute.name === this.name
      );
    });
  }

  /**
   * Is this attribute a pk or sk? If so, it's required in all operations
   */
  public get isRequiredAccessPatternKey(): boolean {
    return this.isPartitionKey || this.isSortKey;
  }

  /**
   * All other access pattern keys that are not the pk or sk are optional.
   * Is Access Pattern Key
   * Is not the SK or SK
   */
  public get isOptionalAccessPatternKey(): boolean {
    return !this.isRequiredAccessPatternKey;
  }

  /*****************************************************************************
   *
   *  PUBLIC vs PRIVATE
   *
   ****************************************************************************/

  public get isPublic(): boolean {
    return this.options.isPublic;
  }

  public get isPrivate(): boolean {
    return !this.isPublic;
  }

  /*****************************************************************************
   *
   *  ACCESS PATTERN HELPERS
   *
   ****************************************************************************/

  public get isPkComponent(): boolean {
    return this.options.isPkComponent;
  }

  public get isSkComponent(): boolean {
    return this.options.isSkComponent;
  }

  public get isLookupComponent(): boolean {
    return this.options.isLookupComponent;
  }

  public get keyAccessPattern(): AccessPattern {
    return this.resource.keyAccessPattern;
  }

  public get lookupAccessPattern(): AccessPattern {
    return this.resource.lookupAccessPattern;
  }

  /*****************************************************************************
   *
   *  GENERATOR HELPERS
   *
   ****************************************************************************/

  public get isGenerated(): boolean {
    return this.options.generator !== AttributeGenerator.NONE;
  }

  public get generator(): ValueOf<typeof AttributeGenerator> {
    return this.options.generator;
  }

  public get isGeneratedOnCreate(): boolean {
    return this.options.isGeneratedOnCreate;
  }

  public get isGeneratedOnRead(): boolean {
    return this.options.isGeneratedOnRead;
  }

  public get isGeneratedOnUpdate(): boolean {
    return this.options.isGeneratedOnUpdate;
  }

  public get isGeneratedOnDelete(): boolean {
    return this.options.isGeneratedOnDelete;
  }

  public get isData(): boolean {
    return this.options.generator === AttributeGenerator.NONE;
  }

  public get isAutoIncrementGenerator(): boolean {
    return (
      this.isGenerated && this.generator === AttributeGenerator.AUTO_INCREMENT
    );
  }

  public get isGuidGenerator(): boolean {
    return this.isGenerated && this.generator === AttributeGenerator.GUID;
  }

  public get isDateTimeGenerator(): boolean {
    return (
      this.isGenerated &&
      this.generator === AttributeGenerator.CURRENT_DATE_TIME_STAMP
    );
  }

  public get isTypeGenerator(): boolean {
    return this.isGenerated && this.generator === AttributeGenerator.TYPE;
  }

  public get isVersionGenerator(): boolean {
    return this.isGenerated && this.generator === AttributeGenerator.VERSION;
  }

  public get isComposableGenerator(): boolean {
    return (
      this.isGenerated && this.generator === AttributeGenerator.COMPOSITION
    );
  }

  /*****************************************************************************
   *
   *  UPSTREAM HELPERS
   *
   ****************************************************************************/

  public get service(): Service {
    return this.resource.service;
  }
}
