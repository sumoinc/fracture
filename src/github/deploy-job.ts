import { Component } from "projen";
import {
  Job,
  JobCallingReusableWorkflow,
  JobPermission,
} from "projen/lib/github/workflows-model";
import { Fracture } from "../core";

export interface DeployJobOptions {
  /**
   * Job ID for the deploy job.
   *
   * @default deploy
   */
  jobId: string;
  /**
   * The command to run to deploy the code.
   *
   * @default "npx projen deploy"
   */
  deployCommand: string;
}

export class DeployJob extends Component {
  /**
   * The job definition, suitable for adding to a workflow.
   */
  public readonly job: Job | JobCallingReusableWorkflow;
  /**
   * The Job Id for the deploy job
   *
   * @default deploy
   */
  public readonly jobId: string;
  /**
   * The command to run to deploy the code.
   *
   * @default "npx projen deploy"
   */
  public readonly deployCommand: string;

  constructor(fracture: Fracture, options: Partial<DeployJobOptions> = {}) {
    super(fracture);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.jobId = options.jobId ?? "deploy";
    this.deployCommand = options.deployCommand ?? "npx projen deploy";

    /***************************************************************************
     * Create job definition
     **************************************************************************/

    this.job = {
      runsOn: ["ubuntu-latest"],
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: [
        {
          name: "Deploy",
          run: this.deployCommand,
        },
      ],
    };
  }
}
