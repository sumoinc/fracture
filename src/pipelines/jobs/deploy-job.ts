import { paramCase } from "change-case";
import { Task } from "projen";
import { Job, JobPermission } from "projen/lib/github/workflows-model";
import { SetOptional, ValueOf } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { REGION_IDENTITIER } from "../../core";
import { renderDeploySteps } from "../steps/deploy-steps";
import { renderDownloadArtifactSteps } from "../steps/download-artifact-steps";
import { renderSetupNodeSteps } from "../steps/setup-node-steps";
import { Workflow } from "../workflow";

export interface DeployJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPattern?: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs?: Array<string>;

  /**
   * Projen Task that runs the deploy
   */
  readonly deployTask: Task;

  /**
   * AWS OIDC Credentials, if deploying to AWS
   */
  readonly awsOidcCredentials?: {
    roleToAssume: string;
    awsRegion: ValueOf<typeof REGION_IDENTITIER>;
    roleDurationSeconds: number;
  };
}

export class DeployJob extends WorkflowJob {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPattern: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs: Array<string>;

  /**
   * Projen Task that runs the deploy
   */
  readonly deployTask: Task;

  /**
   * AWS OIDC Credentials, if deploying to AWS
   */
  readonly awsOidcCredentials?: {
    roleToAssume: string;
    awsRegion: ValueOf<typeof REGION_IDENTITIER>;
    roleDurationSeconds: number;
  };

  constructor(public readonly workflow: Workflow, options: DeployJobOptions) {
    const branchPattern = options.branchPattern ?? "main";
    const name = options.name ?? `Deploy ${branchPattern}`;
    const jobId = options.jobId ?? paramCase(name);

    super(workflow, {
      jobId,
      name,
      ...options,
    });

    // defaults
    this.branchPattern = branchPattern;
    this.needs = [workflow.buildJob.jobId, ...(options.needs ?? [])];
    this.deployTask = options.deployTask;
    this.awsOidcCredentials = options.awsOidcCredentials;

    // make sure build stores these
    workflow.buildJob.artifactDirectories.push(...this.artifactDirectories);
  }

  dependsOn = (job: WorkflowJob): void => {
    this.needs.push(job.jobId);
  };

  render = (): Job => {
    return {
      name: this.name,
      runsOn: this.workflow.defaultRunners,
      needs: this.needs,
      if: `startsWith( github.ref, 'refs/heads/${this.branchPattern}' )`,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.READ,
        ...(this.awsOidcCredentials
          ? {
              idToken: JobPermission.WRITE,
            }
          : {}),
      },
      steps: [
        ...renderDownloadArtifactSteps(this),
        ...renderSetupNodeSteps(this.workflow),
        ...renderDeploySteps(this),
      ],
    };
  };
}
