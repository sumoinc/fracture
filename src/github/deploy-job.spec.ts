import { DeployJob } from "./deploy-job";
import { Fracture } from "../core";

let fracture: Fracture;

beforeEach(() => {
  fracture = new Fracture();
});

test("Default - basic deploy job", () => {
  const buildJob = new DeployJob(fracture);

  expect(buildJob.job).toMatchSnapshot();
  // console.log(JSON.stringify(buildJob.job, null, 2));
});
