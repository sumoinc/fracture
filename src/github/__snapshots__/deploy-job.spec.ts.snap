// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default - basic deploy job 1`] = `
Object {
  "env": Object {
    "CI": "true",
  },
  "permissions": Object {
    "contents": "write",
  },
  "runsOn": Array [
    "ubuntu-latest",
  ],
  "steps": Array [
    Object {
      "name": "Deploy",
      "run": "npx projen deploy",
    },
  ],
}
`;
